add `corepack project install` command

mmkal opened the pull request:
Closes #505

This adds a `corepack project install` command, which is roughly equivalent to

- `corepack enable && npm install` for npm projects
- `corepack enable && pnpm install` for pnpm projects
- `corepack enable && yarn install` for yarn projects

I was able to use it successfully by just running `yarn build` locally, then I could do `node dist/corepack.js project install` to verify it works on the corepack repo itself, and `node ../corepack/dist/corepack.js project install` from various git repos next door.

I wanted to open this up early to get feedback in case the direction isn't considered correct. If it's looking good, I will add tests along similar lines to the existing ones.

Some follow-ons (which I think could be implemented as separate PRs, but could be persuaded to roll them into this one)

- Common CLI args that all the package managers support:
   - `--frozen-lockfile` and `--no-frozen-lockfile`
   - `--lockfile-only`
   - `--no-lockfile`
   - `--production`
   - `--dev`
   - `--offline`
   - `--force`
- Other subcommands:
   - `corepack project add left-pad`
   - `corepack project remove left-pad`

CC @styfle and @aduh95 since you commented on the issue.

styfle commented on sources/commands/Base.ts:
```diff
@@ -32,13 +32,19 @@ export abstract class BaseCommand extends Command<Context> {
+    await this.installLocalPackageManager(info, previousPackageManager);
+  }
+
+  async installLocalPackageManager(info: PreparedPackageManagerInfo, previousPackageManager?: string) {
```
What's the reason for this new function? Or rather, why couldn't you use the existing `setAndInstallLocalPackageManager()` instead?

mmkal replied:
`setAndInstallLocalPackageManager` did this:

<img width="1118" alt="image" src="https://github.com/user-attachments/assets/08af8ff4-9650-4519-970b-412cb11ce43f">

Which is not something I would expect from this command.

`setAndInstallLocalPackageManager` is for the `use` and `up` commands which are supposed to modify your package.json

styfle reviewed (commented):
Looks good to me, just needs tests 👍

mmkal commented:
> Looks good to me, just needs tests 👍

Added! Let me know if you think there's more I should check for - there's not much actual new-new code here so hopefully we're good.

Also LMK if you have thoughts on the follow-ups I listed or if you think there should be others. I could get started on some of them.

styfle commented on sources/commands/Project.ts:
```diff
@@ -0,0 +1,46 @@
+
+      Then, it will run the install command of the specified package manager. If no package manager is specified, it will default to NPM.
+
+      Tt will locate the install directory by running the equivalent of \`which corepack\`.
```
```suggestion
      It will locate the install directory by running the equivalent of \`which corepack\`.
```

styfle reviewed (approved):
LGTM 🎉

aduh95 commented:
Can you add some documentation please?

mmkal commented:
@aduh95 added some in [1510e37](https://github.com/nodejs/corepack/pull/551/commits/1510e374dbc004b84ae3e261f9cb92273d028d11). Let me know if you think more is needed. As mentioned in the PR body, I am happy to do some followups too. I just want to keep it uncontroversial for now to get this base-level functionality in.

anonrig reviewed (approved):
LGTM with more documentation

SuperchupuDev commented on sources/commands/Project.ts:
```diff
@@ -0,0 +1,46 @@
+    details: `
+      When run, this command will check whether the shims for the specified package managers can be found with the correct values inside the install directory. If not, or if they don't exist, they will be created.
+
+      Then, it will run the install command of the specified package manager. If no package manager is specified, it will default to NPM.
```
```suggestion
      Then, it will run the install command of the specified package manager. If no package manager is specified, it will default to npm.
```

SuperchupuDev reviewed (commented):
npm is always lowercase https://npmjs.com/package/npm#faq-on-branding

mmkal commented:
Fixed the tests - they had been working at some point but seemed to have started failing downloading the versions of npm/pnpm/yarn I'd specified, so I just changed them to match versions that are used elsewhere in tests.

Also fixed the NPM->npm casing issue.

@anonrig if you could give another CI approval (I think) this is good to go!

mmkal commented:
@styfle @anonrig do you know who I should ping to get the workflow/updates re-approved? It is passing for me locally but I would love to avoid this getting conflicted in case there are updates needed.

styfle commented:
@aduh95

aduh95 commented:
It looks like there are some linting errors to address.

xtian commented on tests/Project.test.ts:
```diff
@@ -0,0 +1,81 @@
+
+        const dir = await xfs.readdirPromise(cwd);
+        expect(dir).toContain(`package-lock.json`);
+        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true)
```
```suggestion
        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true);
```

xtian commented on tests/Project.test.ts:
```diff
@@ -0,0 +1,81 @@
+
+        const dir = await xfs.readdirPromise(cwd);
+        expect(dir).toContain(`pnpm-lock.yaml`);
+        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true)
```
```suggestion
        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true);
```

xtian commented on tests/Project.test.ts:
```diff
@@ -0,0 +1,81 @@
+
+        const dir = await xfs.readdirPromise(cwd);
+        expect(dir).toContain(`yarn.lock`);
+        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true)
```
```suggestion
        expect(xfs.existsSync(ppath.join(cwd, `node_modules/ms/package.json`))).toBe(true);
```

mmkal commented:
@aduh95 @xtian @anonrig sorry about that, looks like my IDE stopped autoformatting somehow. Now fixed, should be good to go (for real this time!).

mmkal commented:
@aduh95 @xtian @anonrig 👋 , just checking in on this, looks like the code change is approved but the workflow run hasn't been. Could someone click the "approve and run" button so I can make any necessary changes or merge?

spanishpear commented:
@aduh95 @xtian @anonrig sorry for the ping - pretty keen to get this feature - is it possible for someone to run the workflows? Thanks! ❤️
